api/src/wfl/module/covid.clj

```diff
@@ -306,6 +232,139 @@
+  "Throw or return the table from `dataset`"
+  [dataset]
+  (let [[result & more :as all] (-> dataset
+                                    (get-in [:schema :tables])
+                                    (->> (filter (comp #{dataset-table-name} :name))))]
+    (when-not result
+      (throw (ex-info "No table with name" {:name dataset-table-name :dataset dataset})))
+    (when result
+      (throw-unless-column-exists dataset result))
+    result))
+
+(defn verify-source!
+  "Verify that the `dataset` exists and that the WFL has the necessary permissions to read it"
+  [{:keys [name dataset] :as source}]
+  (when-not (some? dataset)
+    (throw (ex-info "Source is nil" {:source source})))
+  (when-not (= (:name source) tdr-source-name)
+    (throw (ex-info "Unknown Source" {:source source})))
+  (-> (datarepo/dataset (:dataset source))
+      (throw-unless-table-exists)))
+
+(defn verify-executor!
+  "Verify the method-configuration exists."
+  [{:keys [name method_configuration] :as executor}]
+  (when-not (= (:name executor) terra-executor-name)
+    (throw (ex-info "Unknown Executor" {:executor executor})))
+  (when-not (:method_configuration executor)
+    (throw (ex-info "Unknown Method Configuration" {:executor executor}))))
+
+(defn verify-sink!
+  "Verify that the WFL has access to both firecloud and the `workspace`."
+  [{:keys [name workspace] :as sink}]
+  (when-not (= (:name sink) terra-workspace-sink-name)
+    (throw (ex-info "Unknown Sink" {:sink sink})))
+  (try
+    (firecloud/get-workspace workspace)
+    (catch Throwable t
+      (throw (ex-info "Cannot access the workspace" {:workspace workspace
+                                                     :cause     (.getMessage t)})))))
+
+(defn ^:private add-continuous-workload-record
+  "Use `tx` and workload `id` to create a \"ContinuousWorkload\" instance and
+  return the ID of the ContinuousWorkload."
+  [tx id {:keys [source sink executor] :as _request}]
+  (let [set-details "UPDATE
+                         ContinuousWorkload
+                     SET
+                         source_type   = ?::source,
+                         executor_type = ?::executor,
+                         sink_type     = ?::sink
+                     WHERE
+                         id = ? "
+        src-exc-snk [(create-source! tx id source)
+                     (create-executor! tx id executor)
+                     (create-sink! tx id sink)]
+        items       (->> (map second src-exc-snk)
+                         (zipmap [:source_items :executor_items :sink_items])
+                         (jdbc/insert! tx :ContinuousWorkload)
+                         first
+                         :id)]
+    (jdbc/execute! tx (concat [set-details] (map first src-exc-snk) [items]))
+    items))
+
+(defn create-covid-workload [tx request]
+  (verify-source! (get-in request [:source]))
+  (verify-executor! (get-in request [:executor]))
+  (verify-sink! (get-in request [:sink]))
```

Comment:
Can you make these multi-methods and implement for source/sink/executor.
Perhaps call the multimethods `throw-when-malformed-NAME-request`, for source, sink, executor.

Reply:
Done